parser/src/string.rs

```diff
@@ -0,0 +1,189 @@
+use crate::ast::{Constant, Expr, ExprKind};
+use crate::error::{LexicalError, LexicalErrorType};
+
+pub fn parse_implicit_concatenation(values: Vec<Expr>) -> Result<Expr, LexicalError> {
+    Ok(if values.len() > 1 {
+        // As in CPython, use the kind of the first Expression.
+        let kind = if let ExprKind::Constant { kind, .. } = &values[0].node {
+            kind.clone()
+        } else {
+            None
+        };
+
+        // Preserve the initial location.
+        let location = values[0].location;
+
+        if values
+            .iter()
+            .all(|value| matches!(value.node, ExprKind::Constant { .. }))
```

Comment:
```suggestion
        let has_fstring = values
            .iter()
            .any(|value| !matches!(value.node, ExprKind::Constant { .. }));
```

Reply:
I think I still need this `has_fstring` flag and logic, even after the refactor, since I need this at the end:

```rust
if has_fstring {
    Expr::new(initial_location, ExprKind::JoinedStr { values: deduped })
} else {
    deduped
        .into_iter()
        .exactly_one()
        .expect("String must be concatenated to a single element.")
}
```

...but correct me if I'm wrong.

Reply:
(In theory, if `values` at the end is just a single `Expr`, and it's a constant, we _could_ just return that -- but CPython seems to return a `JoinedStr` with one constant child, so we'd be deviating from their behavior IIUC.)

Reply:
Oh, yep. Sorry for poor explanation. we need has_fstring, but not by iteration for its own. we can manage a mutable variable to update when it hits fstring branch later.